### ISSUE TITLE
Unit-tests with mockup-db and asyncio motor client

### ISSUE THREAD

**Issue by kob22** [ajdavis/mongo-mockup-db#39]
Hi, I am using fastAPI and mongodb with asyncio motor client. I want to do mongodb mockup for unittests. I tried solution from https://github.com/ajdavis/mongo-mockup-db/issues/25 but it doesn't work.

Python 3.8.5 
motor 2.3.0
pymongo 3.11.2

```@pytest.mark.asyncio
async def test_create_document_in_DB():
    document_example = internal.Document(...)
    server = MockupDB(verbose=True, auto_ismaster=True)
    server.run()

    client = AsyncIOMotorClient(server.uri)
    loop = asyncio.get_event_loop()
    with going(loop.run_until_complete, document.create_document(client, document_example)) as future:
        request = server.receives(OpMsg('insert', 'documents'))
        request.ok()

    # Get the return value of create(), which was passed back through
    # run_until_complete() and going().
    result = future()
    # Just a pymongo.results.InsertOneResult.
    print(result)
```

**Comment by ajdavis** [comment]
Hi, can you describe specifically what's going wrong? Does this code hang, or is there an error message?

**Comment by kob22** [comment]
I am using pytest with @pytest.mark.asyncio and there was a problem with coroutine 'create_docoument' was never awaited and "This event loop is already running" because of asyncio.

So I managed this and this works I think:
```

@pytest.fixture
def server():
    server = MockupDB(verbose=True, auto_ismaster=True)
    server.run()
    yield server
    server.stop()


def test_create_document_in_DB(server):
    document_example = internal.Document(...)

    client = AsyncIOMotorClient(server.uri)
    loop = asyncio.get_event_loop()

    with going(
        loop.run_until_complete, document.create_document(client, document_example)
    ) as future:
        request = server.receives(OpMsg("insert", "documents"))
        request.ok()
```

but it is possible to work with asyncio?
```
@pytest.fixture
def server():
    server = MockupDB(verbose=True, auto_ismaster=True)
    server.run()
    yield server
    server.stop()


@pytest.mark.asyncio
async def test_create_document_in_DB(server):
    document_example = internal.Document(... )

    client = AsyncIOMotorClient(server.uri)
    await document.create_document(client, document_example)
    request = server.receives(OpMsg("insert", "documents"))
    request.ok()
```

**Comment by ajdavis** [comment]
Hi, thanks for providing those details. In your second code example, I think that your call to `await document.create_document(client, document_example)` must time out, because you don't tell MockupDB how to respond to the "insert" message. Perhaps something like this would work?

```
@pytest.mark.asyncio
async def test_create_document_in_DB(server):
    document_example = internal.Document(... )

    client = AsyncIOMotorClient(server.uri)
    # Launch create_document and let it run in parallel with this function.
    task = asyncio.create_task(document.create_document(client, document_example))
    request = server.receives(OpMsg("insert", "documents"))
    request.ok()
    # Now that we've responded to "insert", the task should complete soon.
    result = await task
```

**Comment by kob22** [comment]
Thanks for help. Unfortunately doesn't work.

```
AssertionError: expected to receive OpMsg({"insert": "documents"}), got nothing
```
But I see in log that: 
```
37382   OpMsg({"insert": "documents", "ordered": true, "$db": "tests", (...........)
```

I don't know why this happend. 
A one more error:
```
Task was destroyed but it is pending!
task: <Task pending name='Task-2' coro=<create_document() running at (......)
```

**Comment by ajdavis** [comment]
Sorry, I missed something important. When my previous code executes this line:
```
task = asyncio.create_task(document.create_document(client, document_example))
```
...the task doesn't start running until the next iteration of the asyncio event loop. However, this line:
```
request = server.receives(OpMsg("insert", "documents"))
```
... causes the `test_create_document_in_DB` coroutine to block. In asyncio, if a coroutine blocks, then the whole event loop is blocked and it never proceeds to the next iteration, hence never schedules the task.

(When `server.receives` times out, the coroutine terminates with an exception and the event loop resumes. As the test fails and the loop shuts down, it *does* run the "insert" task, but it's too late.)

Anyway, you need defer `server.receives` to a thread so the loop can run while you wait for it:
```
from concurrent.futures.thread import ThreadPoolExecutor

import pytest
from motor.motor_asyncio import AsyncIOMotorClient
from mockupdb import MockupDB, OpMsg


@pytest.fixture
def server():
    server = MockupDB(verbose=True, auto_ismaster=True)
    server.run()
    yield server
    server.stop()


@pytest.fixture
def executor():
    e = ThreadPoolExecutor()
    yield e
    e.shutdown()


async def create_document(client, document):
    return (await client.db.documents.insert_one(document))


@pytest.mark.asyncio
async def test_create_document_in_DB(event_loop, server, executor):
    document_example = {'x': 1}
    client = AsyncIOMotorClient(server.uri, io_loop=event_loop)
    # Launch create_document and let it run in parallel with this function.
    future = event_loop.create_task(create_document(client, document_example))
    request = await event_loop.run_in_executor(
        executor, server.receives, OpMsg("insert", "documents"))
    request.ok()
    # Now that we've responded to "insert", the task should complete soon.
    result = await future
    print(f'result: {result}')
```
This is inconvenient, hard to get right, and hard to understand. I think MockupDB is very hard to use with async code, and you're not the first person to notice that. Someone should make an async wrapper for MockupDB - what Motor is to PyMongo, this wrapper should be to MockupDB. If there existed a class AsyncMockupDB whose `receives` method deferred its work to a thread and returned a Future, it would be much easier to use with asyncio.
